src/tetris/profiler/hardwareController/cameraController.ts

```diff
@@ -0,0 +1,125 @@
+import HardwarePermission from "tetris/profiler/hardwareController/hardwarePermission";
+
+export default class CameraController {
+
+	//region public members
+	public get permissionState(): HardwarePermission {
+		return this._permissionState;
+	}
+	//endregion
+
+	//region public methods
+	public startVideoStream(): Promise<MediaStream> {
+		if (this.permissionState === HardwarePermission.denied) {
+			return Promise.reject(new Error('No Permissions granted'));
+		}
+		return new Promise( (resolve, reject) => {
+			navigator.mediaDevices.getUserMedia({video: true})
+				.then((stream: MediaStream) => {
+					this._permissionState = HardwarePermission.granted;
+					this._videoElement.classList.add('visible');
+					this._videoElement.srcObject = stream;
+					this._videoStream = stream;
+					this._videoElement.play().then(() => resolve(stream));
+				})
+				.catch(reason => {
+					this._permissionState = HardwarePermission.denied;
+					console.log('Can not start video stream. Message: ' + reason);
+					reject(new Error('Can not start video stream'));
+				});
+		});
+	}
+
+	public stopVideoStream(): void {
+		if (!this._videoStream) {
+			return;
+		}
+		this._videoElement.pause();
+		this._videoStream.stop();
+		this._videoElement.src = null;
+	}
+
+	public takeSnapshot(): Promise<string> {
+		if (this._videoStream) {
+			return Promise.resolve(this._takeSnapshot());
+		}
+		return new Promise((resolve, reject) => {
```

Comment:
Same issue as in line 16. Use `async` and `await` instead.

Reply:
Interesting idea to replace `then()` by `await`/`async`. It indeed improves the nesting level. But I see a big issue regarding our style guideline to type as much as possible.
By using await the returned value can either be the resolved value or the rejected value. In our case these two types might be different (e.g.: resolved => string, rejected => Error). As a result, it is not guaranteed to have an explicitly typed result. 
This would on the one hand complicate our code (since we have to deal with different possible types) and also would violate our style guide. Of course we could handle it individual but this would be inconsistent. 
Therefore, I would suggest to stay with the current implementation.
What do you think?

Reply:
You can just swap the `.catch()` function, against a `try / catch` statement using the `async` call inside. This solves the issue of error handling while preserving simple return types. In this particular case the function could look like this:
Explicit catch and throw:
```
	public async takeSnapshot(): Promise<string> {
		if (this._videoStream) {
			return this._takeSnapshot();
		}
		try {
			await this.startVideoStream();
			return this._takeSnapshot();
		} catch (error) {
			throw error;
		}
	}
```

Since the error is just propagated upwards you can just:
```
	public async takeSnapshot(): Promise<string> {
		if (this._videoStream) {
			return this._takeSnapshot();
		}
		await this.startVideoStream();
		return this._takeSnapshot();
	}
```
which will return an rejected Promise as before, if `this.startVideoStream()` returns an rejected promise.